Add new CLI for PG drop counters part of counterpoll utility

ayurkiv-nvda opened the pull request:
Enhance counterpoll utility to support PG drop counters. 
Add new CLI to enable/disable it.

Signed-off-by: Andriy Yurkiv <ayurkiv@nvidia.com>

**depends on:**
https://github.com/Azure/sonic-buildimage/pull/6444
https://github.com/Azure/sonic-swss/pull/1600

**merge priority:**
https://github.com/Azure/sonic-swss/pull/1600
https://github.com/Azure/sonic-utilities/pull/1355
https://github.com/Azure/sonic-buildimage/pull/6444

<!--
Please make sure you've read and understood our contributing guidelines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "closes #xxxx",
"fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
issue when the PR is merged.

If you are adding/modifying/removing any command or utility script, please also
make sure to add/modify/remove any unit tests from the tests
directory as appropriate.

If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
subcommand, or you are adding a new subcommand, please make sure you also
update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
your changes.

Please provide the following information:
-->

**- What I did**
Added new option for "counterpoll" utility

**- How I did it**
Enhance counterpoll utility for SAI_INGRESS_PRIORITY_GROUP_STAT_DROPPED_PACKETS 
Add new CLI command and extend the show command.

**- How to verify it**
To test it use QoS test. run the pfcxontest and Pfcxofftest, drop can be triggered in either of the test.
It runs in ptf32 topo and requires RPC image.
```
admin@arc-switch1041:~$ counterpoll pg-drop enable  --> to enable the new counter
admin@arc-switch1041:~$ counterpoll show  --> check new INGRESS_PG_STAT_DROP counter status
```
Check counters
```
admin@arc-switch1041:~$ redis-cli -n 2
127.0.0.1:6379[2]> HGETALL COUNTERS:oid:0x1a000000000062
1) "SAI_INGRESS_PRIORITY_GROUP_STAT_XOFF_ROOM_WATERMARK_BYTES"
2) "0"
3) "SAI_INGRESS_PRIORITY_GROUP_STAT_SHARED_WATERMARK_BYTES"
4) "0"
5) "SAI_INGRESS_PRIORITY_GROUP_STAT_DROPPED_PACKETS"
6) "0"

```

**- Previous command output (if the output of a command-line utility has changed)**

```
admin@arc-switch1041:~$ counterpoll show
Type                        Interval (in ms)    Status
--------------------------  ------------------  --------
QUEUE_STAT                  default (10000)     enable
PORT_STAT                   default (1000)      enable
PORT_BUFFER_DROP            default (60000)     enable
RIF_STAT                    default (1000)      enable
QUEUE_WATERMARK_STAT        default (10000)     enable
PG_WATERMARK_STAT           default (10000)     enable
BUFFER_POOL_WATERMARK_STAT  default (10000)     enable
```

```
admin@arc-switch1041:~$ counterpoll
.......
Commands:
  config-db         Config DB counter commands
  port              Queue counter commands
  port-buffer-drop  Port buffer drop counter commands
  queue             Queue counter commands
  rif               RIF counter commands
  show              Show the counter configuration
  watermark         Watermark counter commands
```




**- New command output (if the output of a command-line utility has changed)**
```

admin@arc-switch1041:~$ counterpoll show
Type                        Interval (in ms)    Status
--------------------------  ------------------  --------
QUEUE_STAT                  default (10000)     enable
PORT_STAT                   default (1000)      enable
PORT_BUFFER_DROP            default (60000)     enable
RIF_STAT                    default (1000)      enable
QUEUE_WATERMARK_STAT        default (10000)     enable
PG_WATERMARK_STAT           default (10000)     enable
PG_DROP_STAT                default (10000)     enable         <--- new line
BUFFER_POOL_WATERMARK_STAT  default (10000)     enable

```


```
admin@arc-switch1041:~$ counterpoll
Usage: counterpoll [OPTIONS] COMMAND [ARGS]...
...............
Commands:
  config-db         Config DB counter commands
  pg-drop           Ingress PG drop counter commands     <--- new command
  port              Queue counter commands
  port-buffer-drop  Port buffer drop counter commands
  queue             Queue counter commands
  rif               RIF counter commands
  show              Show the counter configuration
  watermark         Watermark counter commands
```

ayurkiv-nvda commented:
retest this please

volodymyrsamotiy commented:
@neethajohn, could you you please review?

liat-grozovik commented:
@neethajohn if this is needed for 201911 please add the proper label

liat-grozovik commented:
@ayurkiv-nvda the PR description is missing some information. I have modified it a little bit but I believe the 'how to test' is should be extended with how to cause the drop counter to be incremented and then check validity.

You are referring to the show command for the counters which is great but should you verify the reset counter command as well?

Please all add content to the section of 'New command output (if the output of a command-line utility has changed)' as you do have a new command.

neethajohn commented:
To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.

neethajohn commented on tests/counterpoll_test.py:
```diff
@@ -54,6 +55,20 @@ def test_port_buffer_drop_interval_too_short(self):
+        runner = CliRunner()
+        result = runner.invoke(counterpoll.cli.commands["pg-drop"].commands["interval"], ["10000"])
+        print(result.output)
+        assert result.exit_code == 0
```
Need to include the check if the interval field is changed to the new value in config db

ayurkiv-nvda replied:
Done

neethajohn commented on tests/counterpoll_test.py:
```diff
@@ -76,6 +91,13 @@ def test_update_counter_config_db_status(self, status, _get_config_db_file):
+        runner = CliRunner()
+        result = runner.invoke(counterpoll.cli.commands["pg-drop"].commands[status])
+        print(result.output)
+        assert result.exit_code == 0
```
Please also validate that 'flex_counter_status' is accordingly updated in config db

ayurkiv-nvda replied:
Done

ayurkiv-nvda commented:
> To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.

Currently 'counterpoll' utility provide  few command to work with it.
'counterpoll show' - print all counters info in one table. We have general test for it 
'counterpoll <counter command> enable'
'counterpoll <counter command> disable'
'counterpoll <counter command> interval'

There is no possibility to show only specific counter.
The same with clearing. There is no possibility to clean any counters via counterpoll cli

ayurkiv-nvda commented:
> @ayurkiv-nvda the PR description is missing some information. I have modified it a little bit but I believe the 'how to test' is should be extended with how to cause the drop counter to be incremented and then check validity.
> 
> You are referring to the show command for the counters which is great but should you verify the reset counter command as well?
> 
> Please all add content to the section of 'New command output (if the output of a command-line utility has changed)' as you do have a new command.

added

neethajohn commented:
> > To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.
> 
> Currently 'counterpoll' utility provide few commands to work with it.
> 'counterpoll show' - print all counters info in one table. We already have general test for it
> 'counterpoll < counter type> enable'
> 'counterpoll < counter type> disable'
> 'counterpoll < counter type> interval' - three option for each counter type
> 
> There is no possibility to show only specific counter.
> The same with clearing. There is no possibility to clean any counters via counterpoll cli

You cannot do it through 'counterpoll'.  You need a separate cli to view the drop stats. Look at 'show priority-group' or 'show queue' in show/main.py. The same applies for 'clear'

ayurkiv-nvda commented:
> > > To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.
> > 
> > 
> > Currently 'counterpoll' utility provide few commands to work with it.
> > 'counterpoll show' - print all counters info in one table. We already have general test for it
> > 'counterpoll < counter type> enable'
> > 'counterpoll < counter type> disable'
> > 'counterpoll < counter type> interval' - three option for each counter type
> > There is no possibility to show only specific counter.
> > The same with clearing. There is no possibility to clean any counters via counterpoll cli
> 
> You cannot do it through 'counterpoll'. You need a separate cli to view the drop stats. Look at 'show priority-group' or 'show queue' in show/main.py. The same applies for 'clear'

Can we come up with CLI  for this counters? This was not considered at the planning stage and will take a time to implement.
What about this command for **show**:
 **show priority-group dropped_packets**
```
        Port    PG0    PG1    PG2    PG3    PG4    PG5    PG6    PG7
-----------  -----  -----  -----  -----  -----  -----  -----  -----
  Ethernet0      0      0      0      0      0      0      0      0
  Ethernet4      0      0      0      0      0      0      0      0
  Ethernet8      0      0      0      0      0      0      0      0
 ...............................................................
Ethernet116      0      0      0      0      0      0      0      0
Ethernet120      0      0      0      0      0      0      0      0
Ethernet124      0      0      0      0      0      0      0      0
```
And this for **clear**:
**sonic-clear priority-group dropped_packet**
 ?

neethajohn commented:
> > > > To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.
> > > 
> > > 
> > > Currently 'counterpoll' utility provide few commands to work with it.
> > > 'counterpoll show' - print all counters info in one table. We already have general test for it
> > > 'counterpoll < counter type> enable'
> > > 'counterpoll < counter type> disable'
> > > 'counterpoll < counter type> interval' - three option for each counter type
> > > There is no possibility to show only specific counter.
> > > The same with clearing. There is no possibility to clean any counters via counterpoll cli
> > 
> > 
> > You cannot do it through 'counterpoll'. You need a separate cli to view the drop stats. Look at 'show priority-group' or 'show queue' in show/main.py. The same applies for 'clear'
> 
> Can we come up with CLI for this counters? This was not considered at the planning stage and will take a time to implement.
> What about this command for **show**:
> **show priority-group dropped_packets**
> 
> ```
>         Port    PG0    PG1    PG2    PG3    PG4    PG5    PG6    PG7
> -----------  -----  -----  -----  -----  -----  -----  -----  -----
>   Ethernet0      0      0      0      0      0      0      0      0
>   Ethernet4      0      0      0      0      0      0      0      0
>   Ethernet8      0      0      0      0      0      0      0      0
>  ...............................................................
> Ethernet116      0      0      0      0      0      0      0      0
> Ethernet120      0      0      0      0      0      0      0      0
> Ethernet124      0      0      0      0      0      0      0      0
> ```
> 
> And this for **clear**:
> **sonic-clear priority-group dropped_packet**
> ?

How about
show priority-group drop counters
sonic-clear priority-group drop counters

liat-grozovik commented:
@Neetha John<mailto:Neetha.John@microsoft.com> may I suggest to differ this req from the PR? As this was not in the initial discussion and there is no specific CLI for each type of counter we should add it as part of new PR and have this done.

From: Neetha John <notifications@github.com>
Sent: Tuesday, February 2, 2021 6:37 AM
To: Azure/sonic-utilities <sonic-utilities@noreply.github.com>
Cc: Liat Grozovik <liatg@nvidia.com>; Comment <comment@noreply.github.com>
Subject: Re: [Azure/sonic-utilities] Add new CLI for PG drop counters part of counterpoll utility (#1355)


To add on to Liat's comment, I do not see the 'show'/'clear' command to view/clear the pg drop counters.

Currently 'counterpoll' utility provide few commands to work with it.
'counterpoll show' - print all counters info in one table. We already have general test for it
'counterpoll < counter type> enable'
'counterpoll < counter type> disable'
'counterpoll < counter type> interval' - three option for each counter type
There is no possibility to show only specific counter.
The same with clearing. There is no possibility to clean any counters via counterpoll cli

You cannot do it through 'counterpoll'. You need a separate cli to view the drop stats. Look at 'show priority-group' or 'show queue' in show/main.py. The same applies for 'clear'

Can we come up with CLI for this counters? This was not considered at the planning stage and will take a time to implement.
What about this command for show:
show priority-group dropped_packets

        Port    PG0    PG1    PG2    PG3    PG4    PG5    PG6    PG7

-----------  -----  -----  -----  -----  -----  -----  -----  -----

  Ethernet0      0      0      0      0      0      0      0      0

  Ethernet4      0      0      0      0      0      0      0      0

  Ethernet8      0      0      0      0      0      0      0      0

 ...............................................................

Ethernet116      0      0      0      0      0      0      0      0

Ethernet120      0      0      0      0      0      0      0      0

Ethernet124      0      0      0      0      0      0      0      0

And this for clear:
sonic-clear priority-group dropped_packet
?

How about
show priority-group drop counters
sonic-clear priority-group drop counters

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/Azure/sonic-utilities/pull/1355#issuecomment-771352902>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKTABAYDM5MG65CREPXRVLDS456M7ANCNFSM4WBUB7ZQ>.

neethajohn commented:
/AzurePipelines run

liat-grozovik commented:
@ayurkiv-nvda can you please update submodule pointer so we can move forward with the next PR to be merged?

qiluo-msft commented on counterpoll/main.py:
```diff
@@ -123,6 +124,45 @@ def disable():
+@click.pass_context
+def pg_drop(ctx):
+    """  Ingress PG drop counter commands """
+    ctx.obj = swsssdk.ConfigDBConnector()
```
>swsssdk [](start = 14, length = 7)

swsssdk is deprecately largely in sonic-utilities. Could you adapt the new `ConfigDBConnector` as in latest code?

liat-grozovik replied:
is it also deprecated on 202012? this PR should go to 202012 as well thus need to know if the change should be done on specific branch only
@ayurkiv-nvda  FYI
